### ISSUE TITLE
no IPv6 support

### ISSUE THREAD

**Issue by thomasschaeferm** [warner/magic-wormhole#354]
two hosts, the sending host dualstacked IPv4/IPv6, receiving host IPv6-only
The receiving host doesn't see the direct connection
instead of an transfer I get the following message:
`ERROR: We had a problem connecting to the relay server:
(relay URL was ws://relay.magic-wormhole.io:4000/v1)
No route to host: 101: Network is unreachable.
`
The relay server has no IPv6 address.

**Comment by thomasschaeferm** [comment]
Also the socket for transfer is only ipv4.
`LISTEN   0         50                                           0.0.0.0:45587                                               0.0.0.0:*    `

**Comment by warner** [comment]
Hrm, that wasn't intentional. Oops.

The relay server has an ipv6 address of `2600:3c00::f03c:91ff:fe62:b5b5`. I don't have a v6-only host to test from, could you try a `netcat 2600:3c00::f03c:91ff:fe62:b5b5 4000` and see if it connects? If that works, I'll add an A6 record for `relay.magic-wormhole.io`. I'll also dig up the A6 record for the transit relay too.

The transfer socket is using Twisted's `endpoints.serverFromString('tcp:%d")`, which really ought to be v4+v6, but maybe something is less capable than I'd assumed.

Also, we need to update `find_addresses()`. It just runs `ip addr` and parses the results, and the current parser only understands v4 addresses right now.

**Comment by warner** [comment]
er, make that AAAA records, not A6

**Comment by warner** [comment]
I've added AAAA records for both hosts.

BTW, you can do `wormhole send --no-listen FILENAME` and `wormhole rx --no-listen` to disable the listening port on that side, which is useful when testing all the different combinations. If you use `--no-listen` on both sides, the file transfer is forced to go through the transit relay.

The different things we need to test/fix are:

* 1: client-side support for v6 on the "mailbox"/"rendezvous"/"relay" server connection (the websocket to port 4000 on `relay.magic-wormhole.io`
* 2: AAAA record for `relay.magic-wormhole.io`
* 3: server-side support for WebSocket listening on a v6 socket on `relay.magic-wormhole.io`
* 4: client-side support for v6 on transit connections
* 5: client-side support for listening on a v4+v6-capable socket for the transit connection
* 6: client-side support for enumerating both v4 and v6 addresses on that listening socket
* 7: server-side support for listening on a v6 socket on `transit.magic-wormhole.io`
* 8: AAAA record for `transit.magic-wormhole.io`

I should have just fixed 2 and 8 by adding those DNS records.

For 3, `netstat -atn |grep LISTEN |grep 4000` shows:

```
tcp        0      0 0.0.0.0:4000            0.0.0.0:*               LISTEN
```

which I think means it's only listening on v4 (I see some other services saying e.g. `:::NNN` instead-of/in-addition-to `0.0.0.0:NNN`, so the fact that I don't see any instances of `:::4000` suggests we're not listening on v6. So I think items 3 and 7 are both broken, and require changes to the way we set up that socket. That'll be in https://github.com/warner/magic-wormhole-mailbox-server/blob/0.3.1/src/wormhole_mailbox_server/server_tap.py#L100 and https://github.com/warner/magic-wormhole-transit-relay/blob/0.1.2/src/wormhole_transit_relay/server_tap.py#L34 .

And I bet I use the same code pattern to create the client transit listening socket, in https://github.com/warner/magic-wormhole/blob/0.11.2/src/wormhole/transit.py#L668 , so I imagine 5 is broken too.

I'd like to think that 1 and 4 work, because I'm not doing anything v6-specific there, but who knows. 4 happens at https://github.com/warner/magic-wormhole/blob/13102d3eeded33a4ad749078de9bbca5000ffbe6/src/wormhole/transit.py#L964-L966 , and uses `HostnameEndpoint`, which is supposed to try both AAAA and A records. 1 happens somewhere inside the `autobahn` library, which accepts a URL, so I'd like to think that's already doing as much IPv6 as we need.

So, the action items are:

* fix 3/5/7 by figuring out what sort of server-ish Endpoint is needed to listen on both v4 and v6
* update three places to use those: client transit listening socket, mailbox server, transit relay server
* make new releases of the two servers, deploy to `mailbox.magic-wormhole.io` and `transit.magic-wormhole.io`
* test 1/4 (client-side v6 connections) once the servers can accept v6 connections
* fix 6 by enhancing `find_addresses()` to parse IPv6 addresses in the output of `ip a list`, not just v4

**Comment by warner** [comment]
Ah, an endpoint string of `tcp:NNN:interface=\:\:0` will get us a v4+v6 listening socket. The default must do `interface=127.0.0.1` which results in a v4-only listening socket.

**Comment by warner** [comment]
Nope, that gets a v6-only socket. We have to use two separate endpoints to get both. Also, `interface=127.0.0.1` is wrong, it's probably `interface=0.0.0.0`.

**Comment by warner** [comment]
Ok, doing a single endpoint of `tcp:4000:interface=\:\:` seems to work. On macOS, netstat reports:

```
% netstat -a -n -f inet|grep LISTEN|grep 4000
tcp46      0      0  *.4000                 *.*                    LISTEN
```

and both `nc -4v localhost 4000` and `nc -6v localhost 4000` are able to connect. On linux, netstat reports:

```
$ netstat -atn |grep 4000
tcp6       0      0 :::4000                 :::*                    LISTEN
```

and both 4 and 6 netcats can connect.

(I tried using two endpoints, one without an `interface=` and the other with `interface=\:\:`.. it worked on macOS, but failed with an EADDRINUSE on linux).

**Comment by warner** [comment]
- [x] fix 3 and 7: change endpoint to listen on both v4 and v6
- [ ] fix 5: change client listener endpoint
- [x] make new releases of mailbox server and transit relay server
- [x] deploy new releases to mailbox.magic-wormhole.io and transit.magic-wormhole.io
- [ ] test 1 and 4 with a v6-only client
- [ ] fix 6 by enhancing `find_addresses()` to parse v6 addresses in `ip a list`

**Comment by warner** [comment]
@thomasschaeferm could you give it another try now? I think your v6-only box should be able to get to both servers, so using `--no-listen` on both sides should exercise connectivity to both the mailbox/rendezvous/relay server, and the transit-relay server.

When we fix point 5 and 6, the test will be to run the sender from the v4+v6 host with `wormhole --transit-helper tcp:localhost:4444 send --no-listen FILENAME`, and the receiver from the v6-only host with `wormhole --transit-helper tcp:localhost:4444 rx --no-listen`. That will effectively disable the transit relay, and disable the listener on the sending side, so the only remaining connection possibility is for the v6 receiver to correctly announce it's IPv6 address, and for the sender to connect to it.

I want to be a bit more cautious about fixing 5, because if the client machine is IPv4-only, doing a listening socket on `::` might fail entirely. I want to test this on a machine that has v6 removed entirely (probably by rebuilding the kernel, or maybe there's a sysconfig option), and I may need to have some code that detects whether `::` works or not and fall back to a v4-only listening socket if it doesn't.

(I could be somewhat cavalier about changing the servers to unconditionally use `::`, because if there's a problem it'll happen right away, and there's an easy `--port=tcp:4000` workaround)

**Comment by thomasschaeferm** [comment]
Sorry for let waiting you so long. I did not expect such a fast reaction.
Your transit relay now works in both directions. It also translates fine from ipv4-only hosts to ipv6-only hosts and vice versa. 
So wormhole works. Of course it could be better in finding better/shorter ways ( my dualstack-host is within the same LAN as ipv6-only host)
For further tests I probably need recent master versions instead of the packaged versions.  I will be glad to do further tests, but not today.
Thanks a lot for providing such a useful tool.

**Comment by warner** [comment]
Great, thanks for the update!

I'm going to leave the ticket open until I've landed the client-listener changes too, which ought to help your dual-stack host find the shorter path. I'll ping you here when I have something ready for you to test out.
